File: Dot examples/ASP Book Examples/Book Examples/Chapter30/CS/App_Code/ArticleWithDefault.cs
using System;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace myControls
{

    public class ArticleWithDefault : CompositeControl
    {

        private string _title;
        private string _author;
        private string _contents;

        private ITemplate _itemTemplate;

        public string Title
        {
            get { return _title; }
            set { _title = value; }
        }

        public string Author
        {
            get { return _author; }
            set { _author = value; }
        }

        public string Contents
        {
            get { return _contents; }
            set { _contents = value; }
        }

        [TemplateContainer(typeof(ArticleWithDefault))]
        [PersistenceMode(PersistenceMode.InnerProperty)]
        public ITemplate ItemTemplate
        {
            get { return _itemTemplate; }
            set { _itemTemplate = value; }
        }

        protected override void CreateChildControls()
        {
            if (_itemTemplate == null)
                _itemTemplate = new ArticleDefaultTemplate();
            _itemTemplate.InstantiateIn(this);
        }
    }

    public class ArticleDefaultTemplate : ITemplate
    {
        public void InstantiateIn(Control container)
        {
            Label lblTitle = new Label();
            lblTitle.DataBinding += new EventHandler(lblTitle_DataBinding);

            Label lblAuthor = new Label();
            lblAuthor.DataBinding += new EventHandler(lblAuthor_DataBinding);

            Label lblContents = new Label();
            lblContents.DataBinding += new EventHandler(lblContents_DataBinding);

            container.Controls.Add(lblTitle);
            container.Controls.Add(new LiteralControl("<br />"));
            container.Controls.Add(lblAuthor);
            container.Controls.Add(new LiteralControl("<br />"));
            container.Controls.Add(lblContents);
        }

        void lblTitle_DataBinding(object sender, EventArgs e)
        {
            Label lblTitle = (Label)sender;
            ArticleWithDefault container = (ArticleWithDefault)lblTitle.NamingContainer;
            lblTitle.Text = container.Title;
        }

        void lblAuthor_DataBinding(object sender, EventArgs e)
        {
            Label lblAuthor = (Label)sender;
            ArticleWithDefault container = (ArticleWithDefault)lblAuthor.NamingContainer;
            lblAuthor.Text = container.Author;
        }

        void lblContents_DataBinding(object sender, EventArgs e)
        {
            Label lblContents = (Label)sender;
            ArticleWithDefault container = (ArticleWithDefault)lblContents.NamingContainer;
            lblContents.Text = container.Contents;
        }

    }

}

File: Dot examples/ASP Book Examples/Book Examples/Chapter30/VB/App_Code/ArticleWithDefault.vb
Imports System
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls

Namespace myControls

    Public Class ArticleWithDefault
        Inherits CompositeControl

        Private _title As String
        Private _author As String
        Private _contents As String

        Private _itemTemplate As ITemplate

        Public Property Title() As String
            Get
                Return _title
            End Get
            Set(ByVal Value As String)
                _title = value
            End Set
        End Property

        Public Property Author() As String
            Get
                Return _author
            End Get
            Set(ByVal Value As String)
                _author = value
            End Set
        End Property

        Public Property Contents() As String
            Get
                Return _contents
            End Get
            Set(ByVal Value As String)
                _contents = value
            End Set
        End Property

        <TemplateContainer(GetType(ArticleWithDefault))> _
        <PersistenceMode(PersistenceMode.InnerProperty)> _
        Public Property ItemTemplate() As ITemplate
            Get
                Return _itemTemplate
            End Get
            Set(ByVal Value As ITemplate)
                _itemTemplate = Value
            End Set
        End Property

        Protected Overrides Sub CreateChildControls()
            If _itemTemplate Is Nothing Then
                _itemTemplate = New ArticleDefaultTemplate()
            End If
            _itemTemplate.InstantiateIn(Me)
        End Sub
    End Class

    Public Class ArticleDefaultTemplate
        Implements ITemplate

        Public Sub InstantiateIn(ByVal container As Control) Implements ITemplate.InstantiateIn
            Dim lblTitle As New Label()
            AddHandler lblTitle.DataBinding, AddressOf lblTitle_DataBinding

            Dim lblAuthor As New Label()
            AddHandler lblAuthor.DataBinding, AddressOf lblAuthor_DataBinding

            Dim lblContents As New Label()
            AddHandler lblContents.DataBinding, AddressOf lblContents_DataBinding

            container.Controls.Add(lblTitle)
            container.Controls.Add(New LiteralControl("<br />"))
            container.Controls.Add(lblAuthor)
            container.Controls.Add(New LiteralControl("<br />"))
            container.Controls.Add(lblContents)
        End Sub

        Private Sub lblTitle_DataBinding(ByVal sender As Object, ByVal e As EventArgs)
            Dim lblTitle As Label = CType(sender, Label)
            Dim container As ArticleWithDefault = CType(lblTitle.NamingContainer, ArticleWithDefault)
            lblTitle.Text = container.Title
        End Sub

        Private Sub lblAuthor_DataBinding(ByVal sender As Object, ByVal e As EventArgs)
            Dim lblAuthor As Label = CType(sender, Label)
            Dim container As ArticleWithDefault = CType(lblAuthor.NamingContainer, ArticleWithDefault)
            lblAuthor.Text = container.Author
        End Sub

        Private Sub lblContents_DataBinding(ByVal sender As Object, ByVal e As EventArgs)
            Dim lblContents As Label = CType(sender, Label)
            Dim container As ArticleWithDefault = CType(lblContents.NamingContainer, ArticleWithDefault)
            lblContents.Text = container.Contents
        End Sub

    End Class

End Namespace

